xdsl/transforms/experimental/live_range.py

```diff
@@ -0,0 +1,52 @@
+from dataclasses import dataclass
+from xdsl.ir import Block, OpResult, SSAValue
+from xdsl.utils.exceptions import InvalidIRException
+
+
+@dataclass(init=False, frozen=True)
+class LiveRange:
+    """
+    Represents the live range of an SSA value as half-open interval using the
+    index of an Operation within its parent Block.
+    """
+
+    value: SSAValue
+    start: int
+    end: int
+
+    def __init__(
+        self, value: SSAValue, start: int | None = None, end: int | None = None
+    ) -> None:
+        if start is not None and end is not None:
+            object.__setattr__(self, "value", value)
+            object.__setattr__(self, "start", start)
+            object.__setattr__(self, "end", end)
+        else:
+            if not value.owner:
+                raise InvalidIRException(
+                    "Cannot calculate live range for value not belonging to a block"
+                )
+            if isinstance(value.owner, Block):
+                raise NotImplementedError("Cannot support block arguments")
+
+            owner = value.owner
+
+            if isinstance(value, OpResult) and isinstance(owner.parent, Block):
+                object.__setattr__(self, "value", value)
+                object.__setattr__(
+                    self, "start", owner.parent.get_operation_index(value.owner)
+                )
+
+                end = self.start
+                for use in value.uses:
```

Comment:
If we walk backward through the basic block, the first use that we find starts a live range. In that setting, I am actually curious if we even need a separate live-range data structure or if we could write the algorithm to just walk backwards. Walking backwards through the sequence of instructions makes the non-spilling variant of this algorithm easy:

for op in reverse(bb.ops):
  for operand in op.operands:
      reserve_register_if_not_already_reserved(operand)
  for res in op.results:
      free_register(res)

Reply:
I tried reversing `BlockOps` and it is not supported, so I ended up "caching" the operations in a list.
It felt clunkier unless we want to enhance that API.

Moreover, moving to spill/reload handling and exploring potential splitting (if required) soon.

Reply:
> However, our immediate goal is to be able to allocate the PulpNN kernels and I feel this is the shortest path.

Are we in a rush? I feel taking some time to explore new ways to do register allocation that are enabled by MLIR could be nice.

Reply:
It would be nice to add `walk_reverse` everywhere we have `walk`, to do this well.

Reply:
One thing that partial register allocation would let us do is some constraint-based iterative system. Where we build a graph of all ssa values that are alive at the same time (or might be alive at the same time), and in a loop assign one of the unused registers to the most constricted unassigned registers. This approach wouldn't rely on indices at all, just a single walk forwards.

Reply:
https://github.com/xdslproject/xdsl/pull/1115 adds walk_reverse().

Reply:
https://github.com/xdslproject/xdsl/pull/1115 is now merged thanks to @superlopuh's quick review.